File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-asset-5-4/CurveInstrument.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CurveInstrument : IdentifiedAsset
    {
        public CurveInstrument(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode currencyNode = xmlNode.SelectSingleNode("currency");
            
            if (currencyNode != null)
            {
                if (currencyNode.Attributes["href"] != null || currencyNode.Attributes["id"] != null) 
                {
                    if (currencyNode.Attributes["id"] != null) 
                    {
                        currencyIDRef_ = currencyNode.Attributes["id"].Value;
                        Currency ob = new Currency(currencyNode);
                        IDManager.SetID(currencyIDRef_, ob);
                    }
                    else if (currencyNode.Attributes["href"] != null)
                    {
                        currencyIDRef_ = currencyNode.Attributes["href"].Value;
                    }
                    else
                    {
                        currency_ = new Currency(currencyNode);
                    }
                }
                else
                {
                    currency_ = new Currency(currencyNode);
                }
            }
            
        
            XmlNode exchangeIdNode = xmlNode.SelectSingleNode("exchangeId");
            
            if (exchangeIdNode != null)
            {
                if (exchangeIdNode.Attributes["href"] != null || exchangeIdNode.Attributes["id"] != null) 
                {
                    if (exchangeIdNode.Attributes["id"] != null) 
                    {
                        exchangeIdIDRef_ = exchangeIdNode.Attributes["id"].Value;
                        ExchangeId ob = new ExchangeId(exchangeIdNode);
                        IDManager.SetID(exchangeIdIDRef_, ob);
                    }
                    else if (exchangeIdNode.Attributes["href"] != null)
                    {
                        exchangeIdIDRef_ = exchangeIdNode.Attributes["href"].Value;
                    }
                    else
                    {
                        exchangeId_ = new ExchangeId(exchangeIdNode);
                    }
                }
                else
                {
                    exchangeId_ = new ExchangeId(exchangeIdNode);
                }
            }
            
        
            XmlNode clearanceSystemNode = xmlNode.SelectSingleNode("clearanceSystem");
            
            if (clearanceSystemNode != null)
            {
                if (clearanceSystemNode.Attributes["href"] != null || clearanceSystemNode.Attributes["id"] != null) 
                {
                    if (clearanceSystemNode.Attributes["id"] != null) 
                    {
                        clearanceSystemIDRef_ = clearanceSystemNode.Attributes["id"].Value;
                        ClearanceSystem ob = new ClearanceSystem(clearanceSystemNode);
                        IDManager.SetID(clearanceSystemIDRef_, ob);
                    }
                    else if (clearanceSystemNode.Attributes["href"] != null)
                    {
                        clearanceSystemIDRef_ = clearanceSystemNode.Attributes["href"].Value;
                    }
                    else
                    {
                        clearanceSystem_ = new ClearanceSystem(clearanceSystemNode);
                    }
                }
                else
                {
                    clearanceSystem_ = new ClearanceSystem(clearanceSystemNode);
                }
            }
            
        
            XmlNode definitionNode = xmlNode.SelectSingleNode("definition");
            
            if (definitionNode != null)
            {
                if (definitionNode.Attributes["href"] != null || definitionNode.Attributes["id"] != null) 
                {
                    if (definitionNode.Attributes["id"] != null) 
                    {
                        definitionIDRef_ = definitionNode.Attributes["id"].Value;
                        ProductReference ob = new ProductReference(definitionNode);
                        IDManager.SetID(definitionIDRef_, ob);
                    }
                    else if (definitionNode.Attributes["href"] != null)
                    {
                        definitionIDRef_ = definitionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        definition_ = new ProductReference(definitionNode);
                    }
                }
                else
                {
                    definition_ = new ProductReference(definitionNode);
                }
            }
            
        
        }
        
    
        #region Currency_
        private Currency currency_;
        public Currency Currency_
        {
            get
            {
                if (this.currency_ != null)
                {
                    return this.currency_; 
                }
                else if (this.currencyIDRef_ != null)
                {
                    currency_ = IDManager.getID(currencyIDRef_) as Currency;
                    return this.currency_; 
                }
                else
                {
                      return this.currency_; 
                }
            }
            set
            {
                if (this.currency_ != value)
                {
                    this.currency_ = value;
                }
            }
        }
        #endregion
        
        public string currencyIDRef_ { get; set; }
        #region ExchangeId_
        private ExchangeId exchangeId_;
        public ExchangeId ExchangeId_
        {
            get
            {
                if (this.exchangeId_ != null)
                {
                    return this.exchangeId_; 
                }
                else if (this.exchangeIdIDRef_ != null)
                {
                    exchangeId_ = IDManager.getID(exchangeIdIDRef_) as ExchangeId;
                    return this.exchangeId_; 
                }
                else
                {
                      return this.exchangeId_; 
                }
            }
            set
            {
                if (this.exchangeId_ != value)
                {
                    this.exchangeId_ = value;
                }
            }
        }
        #endregion
        
        public string exchangeIdIDRef_ { get; set; }
        #region ClearanceSystem_
        private ClearanceSystem clearanceSystem_;
        public ClearanceSystem ClearanceSystem_
        {
            get
            {
                if (this.clearanceSystem_ != null)
                {
                    return this.clearanceSystem_; 
                }
                else if (this.clearanceSystemIDRef_ != null)
                {
                    clearanceSystem_ = IDManager.getID(clearanceSystemIDRef_) as ClearanceSystem;
                    return this.clearanceSystem_; 
                }
                else
                {
                      return this.clearanceSystem_; 
                }
            }
            set
            {
                if (this.clearanceSystem_ != value)
                {
                    this.clearanceSystem_ = value;
                }
            }
        }
        #endregion
        
        public string clearanceSystemIDRef_ { get; set; }
        #region Definition_
        private ProductReference definition_;
        public ProductReference Definition_
        {
            get
            {
                if (this.definition_ != null)
                {
                    return this.definition_; 
                }
                else if (this.definitionIDRef_ != null)
                {
                    definition_ = IDManager.getID(definitionIDRef_) as ProductReference;
                    return this.definition_; 
                }
                else
                {
                      return this.definition_; 
                }
            }
            set
            {
                if (this.definition_ != value)
                {
                    this.definition_ = value;
                }
            }
        }
        #endregion
        
        public string definitionIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-asset-5-4/CurveInstrument.cpp
// CurveInstrument.cpp 
#include "CurveInstrument.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CurveInstrument::CurveInstrument(TiXmlNode* xmlNode)
: IdentifiedAsset(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //currencyNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* currencyNode = xmlNode->FirstChildElement("currency");

   if(currencyNode){currencyIsNull_ = false;}
   else{currencyIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- currencyNode , address : " << currencyNode << std::endl;
   #endif
   if(currencyNode)
   {
      if(currencyNode->Attribute("href") || currencyNode->Attribute("id"))
      {
          if(currencyNode->Attribute("id"))
          {
             currencyIDRef_ = currencyNode->Attribute("id");
             currency_ = boost::shared_ptr<Currency>(new Currency(currencyNode));
             currency_->setID(currencyIDRef_);
             IDManager::instance().setID(currencyIDRef_,currency_);
          }
          else if(currencyNode->Attribute("href")) { currencyIDRef_ = currencyNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { currency_ = boost::shared_ptr<Currency>(new Currency(currencyNode));}
   }

   //exchangeIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* exchangeIdNode = xmlNode->FirstChildElement("exchangeId");

   if(exchangeIdNode){exchangeIdIsNull_ = false;}
   else{exchangeIdIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- exchangeIdNode , address : " << exchangeIdNode << std::endl;
   #endif
   if(exchangeIdNode)
   {
      if(exchangeIdNode->Attribute("href") || exchangeIdNode->Attribute("id"))
      {
          if(exchangeIdNode->Attribute("id"))
          {
             exchangeIdIDRef_ = exchangeIdNode->Attribute("id");
             exchangeId_ = boost::shared_ptr<ExchangeId>(new ExchangeId(exchangeIdNode));
             exchangeId_->setID(exchangeIdIDRef_);
             IDManager::instance().setID(exchangeIdIDRef_,exchangeId_);
          }
          else if(exchangeIdNode->Attribute("href")) { exchangeIdIDRef_ = exchangeIdNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { exchangeId_ = boost::shared_ptr<ExchangeId>(new ExchangeId(exchangeIdNode));}
   }

   //clearanceSystemNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* clearanceSystemNode = xmlNode->FirstChildElement("clearanceSystem");

   if(clearanceSystemNode){clearanceSystemIsNull_ = false;}
   else{clearanceSystemIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- clearanceSystemNode , address : " << clearanceSystemNode << std::endl;
   #endif
   if(clearanceSystemNode)
   {
      if(clearanceSystemNode->Attribute("href") || clearanceSystemNode->Attribute("id"))
      {
          if(clearanceSystemNode->Attribute("id"))
          {
             clearanceSystemIDRef_ = clearanceSystemNode->Attribute("id");
             clearanceSystem_ = boost::shared_ptr<ClearanceSystem>(new ClearanceSystem(clearanceSystemNode));
             clearanceSystem_->setID(clearanceSystemIDRef_);
             IDManager::instance().setID(clearanceSystemIDRef_,clearanceSystem_);
          }
          else if(clearanceSystemNode->Attribute("href")) { clearanceSystemIDRef_ = clearanceSystemNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { clearanceSystem_ = boost::shared_ptr<ClearanceSystem>(new ClearanceSystem(clearanceSystemNode));}
   }

   //definitionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* definitionNode = xmlNode->FirstChildElement("definition");

   if(definitionNode){definitionIsNull_ = false;}
   else{definitionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- definitionNode , address : " << definitionNode << std::endl;
   #endif
   if(definitionNode)
   {
      if(definitionNode->Attribute("href") || definitionNode->Attribute("id"))
      {
          if(definitionNode->Attribute("id"))
          {
             definitionIDRef_ = definitionNode->Attribute("id");
             definition_ = boost::shared_ptr<ProductReference>(new ProductReference(definitionNode));
             definition_->setID(definitionIDRef_);
             IDManager::instance().setID(definitionIDRef_,definition_);
          }
          else if(definitionNode->Attribute("href")) { definitionIDRef_ = definitionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { definition_ = boost::shared_ptr<ProductReference>(new ProductReference(definitionNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<Currency> CurveInstrument::getCurrency()
{
   if(!this->currencyIsNull_){
        if(currencyIDRef_ != ""){
             return boost::shared_static_cast<Currency>(IDManager::instance().getID(currencyIDRef_));
        }else{
             return this->currency_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Currency>();
   }
}
boost::shared_ptr<ExchangeId> CurveInstrument::getExchangeId()
{
   if(!this->exchangeIdIsNull_){
        if(exchangeIdIDRef_ != ""){
             return boost::shared_static_cast<ExchangeId>(IDManager::instance().getID(exchangeIdIDRef_));
        }else{
             return this->exchangeId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ExchangeId>();
   }
}
boost::shared_ptr<ClearanceSystem> CurveInstrument::getClearanceSystem()
{
   if(!this->clearanceSystemIsNull_){
        if(clearanceSystemIDRef_ != ""){
             return boost::shared_static_cast<ClearanceSystem>(IDManager::instance().getID(clearanceSystemIDRef_));
        }else{
             return this->clearanceSystem_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ClearanceSystem>();
   }
}
boost::shared_ptr<ProductReference> CurveInstrument::getDefinition()
{
   if(!this->definitionIsNull_){
        if(definitionIDRef_ != ""){
             return boost::shared_static_cast<ProductReference>(IDManager::instance().getID(definitionIDRef_));
        }else{
             return this->definition_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ProductReference>();
   }
}
}

